packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue.ts

```diff
@@ -0,0 +1,214 @@
+import * as ecs from '@aws-cdk/aws-ecs';
+import * as sns from '@aws-cdk/aws-sns';
+import * as subscription from '@aws-cdk/aws-sns-subscriptions';
+import * as sqs from '@aws-cdk/aws-sqs';
+import * as cdk from '@aws-cdk/core';
+import { Service } from '../service';
+import { Container } from './container';
+import { ContainerMutatingHook, ServiceExtension } from './extension-interfaces';
+
+// Keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * An interface that will be implemented by all the resources that can be subscribed to.
+ */
+export interface ISubscribable {
+  /**
+   * All classes implementing this interface must also implement the `subscribe()` method
+   */
+  subscribe(queue: sqs.IQueue): sqs.IQueue | undefined;
+}
+
+/**
+ * The settings for the Queue extension.
+ */
+export interface QueueExtensionProps {
+  /**
+   * The list of subscriptions for this service.
+   *
+   * @default none
+   */
+  readonly subscriptions?: ISubscribable[];
+
+  /**
+   * The user-provided default queue for this service.
+   *
+   * @default If the `eventsQueue` is not provided, a default SQS Queue is created for the service.
+   */
+  readonly eventsQueue?: sqs.IQueue;
+}
+
+/**
+ * The topic-specific settings for creating the queue subscriptions.
+ */
+export interface TopicSubscriptionProps {
+  /**
+   * The SNS Topic to subscribe to.
+   */
+  readonly topic: sns.ITopic;
+
+  /**
+   * The user-provided queue to subscribe to the given topic.
+   * If the `queue` is not provided, the default `eventsQueue` will subscribe to the given topic.
+   *
+   * @default none
+   */
+  readonly queue?: sqs.IQueue;
+}
+
+/**
+ * The `TopicSubscription` class represents an SNS Topic resource that can be subscribed to by the service queues.
+ */
+export class TopicSubscription implements ISubscribable {
+  public readonly topic: sns.ITopic;
+
+  public readonly queue?: sqs.IQueue;
+
+  constructor(props: TopicSubscriptionProps) {
+    this.topic = props.topic;
+    this.queue = props.queue;
+  }
+
+  /**
+   * This method sets up SNS Topic subscriptions for the SQS queue provided by the user. If a `queue` is not provided,
+   * the default `eventsQueue` subscribes to the given topic.
+   *
+   * @param queue `eventsQueue` of the service
+   * @returns topic-specific SQS Queue, if provided by the user
+   */
+  public subscribe(queue: sqs.IQueue) : sqs.IQueue | undefined {
+    if (this.queue) {
+      this.topic.addSubscription(new subscription.SqsSubscription(this.queue));
+    } else {
+      this.topic.addSubscription(new subscription.SqsSubscription(queue));
+    }
+    return this.queue;
+  }
```

Comment:
From our discussion, what do you think of having `subscribe` instead take as input a `QueueExtension`?

```ts
public subscribe(extension: QueueExtension) {
   let queue = extension.eventsQueue();
   if (this.queue) {
       queue = this.queue;
   }
   this.topic.addSubscription(new subscription.SqsSubscription(queue));
}
```
This way, I think the interface becomes a bit clearer. If a `TopicSubscription` has a `queue` then we will use it, otherwise it will be the `eventsQueue`.

Reply:
Yup, looks neat!